tasmota/tasmota_xsns_sensor/xsns_33_qmc5883l.ino

```diff
@@ -0,0 +1,397 @@
+
+  Table 20. SET/RESET Period Register
+  Addr.     7         6         5         4         3         2         1         0
+  0BH       SET/RESET Perio  [7:0]
+
+*/
+
+#ifdef USE_I2C
+#ifdef USE_QMC5883L
+
+/*********************************************************************************************\
+ * QMC5883L is 3-Axis Digital Compass sensor
+ *
+ * Source: Helge Scheunemann
+ *
+ * I2C Address: 0x0D
+\*********************************************************************************************/
+
+// Define driver ID
+#define XSNS_33                          33
+#define XI2C_71                          71  // See I2CDEVICES.md
+
+/* The default I2C address of this chip */
+#define QMC5883L_ADDR                   0x0D
+
+/* Register numbers */
+#define QMC5883L_X_LSB                  0x00
+#define QMC5883L_X_MSB                  0x01
+#define QMC5883L_Y_LSB                  0x02
+#define QMC5883L_Y_MSB                  0x03
+#define QMC5883L_Z_LSB                  0x04
+#define QMC5883L_Z_MSB                  0x05
+#define QMC5883L_STATUS                 0x06
+#define QMC5883L_TEMP_LSB               0x07
+#define QMC5883L_TEMP_MSB               0x08
+#define QMC5883L_CONFIG                 0x09
+#define QMC5883L_CONFIG2                0x0a
+#define QMC5883L_RESET                  0x0b
+#define QMC5883L_RESERVED               0x0c
+#define QMC5883L_CHIP_ID                0x0d
+
+/* Bit values for the STATUS register */
+#define QMC5883L_STATUS_DRDY            1
+#define QMC5883L_STATUS_OVL             2
+#define QMC5883L_STATUS_DOR             4
+
+/* Oversampling values for the CONFIG register */
+#define QMC5883L_CONFIG_OS512           0b00000000
+#define QMC5883L_CONFIG_OS256           0b01000000
+#define QMC5883L_CONFIG_OS128           0b10000000
+#define QMC5883L_CONFIG_OS64            0b11000000
+
+/* Range values for the CONFIG register */
+#define QMC5883L_CONFIG_2GAUSS          0b00000000
+#define QMC5883L_CONFIG_8GAUSS          0b00010000
+
+/* Rate values for the CONFIG register */
+#define QMC5883L_CONFIG_10HZ            0b00000000
+#define QMC5883L_CONFIG_50HZ            0b00000100
+#define QMC5883L_CONFIG_100HZ           0b00001000
+#define QMC5883L_CONFIG_200HZ           0b00001100
+
+/* Mode values for the CONFIG register */
+#define QMC5883L_CONFIG_STANDBY         0b00000000
+#define QMC5883L_CONFIG_CONT            0b00000001
+
+/* Mode values for the CONFIG2 register */
+#define QMC5883L_CONFIG2_RESET          0b10000000
+
+/* Apparently M_PI isn't available in all environments. */
+#ifndef M_PI
+#define M_PI 3.14159265358979323846264338327950288
+#endif
+
+// data field
+struct {
+int16_t   MX = 0, MY = 0, MZ = 0, HG = 0;
+int16_t   xhigh = 0, yhigh = 0, xlow = 0, ylow = 0;
+int16_t   temp = 0;
+uint16_t  scalar = 0;
+uint8_t   status;
+bool      ready = false;
+uint8_t   i2c_address = QMC5883L_ADDR;
+} QMC5883L;
+
+void writeRegister(uint8_t reg, uint8_t val)
```

Comment:
As these functions are global there is a good chance they are already used somewhere. I propose you dive into the Tasmota I2C functions and use them. See ``support.ino`` and search for ``I2CWrite8`` and pick whatever you need.